src/internal/obj/testsuite.go

```diff
@@ -0,0 +1,161 @@
+package obj
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"io"
+	"io/ioutil"
+	"path"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pacherr"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachhash"
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+	tu "github.com/pachyderm/pachyderm/v2/src/internal/testutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+)
+
+// TestSuite runs tests to ensure the object returned by newClient implements Client
+// newClient should register cleanup of the returned object using testing.T.Cleanup
+// All of the subtests call t.Parallel, but the top level test does not.
+func TestSuite(t *testing.T, newClient func(t testing.TB) Client) {
+	ctx := context.Background()
+	t.Run("TestStorage", func(t *testing.T) {
+		t.Parallel()
+		c := newClient(t)
+		require.NoError(t, TestStorage(ctx, c))
+	})
+
+	t.Run("TestMissingObject", func(t *testing.T) {
+		t.Parallel()
+		client := newClient(t)
+		object := tu.UniqueString("test-missing-object-")
+		requireExists(t, client, object, false)
+
+		err := client.Get(ctx, object, 0, 0, &bytes.Buffer{})
+		require.YesError(t, err)
+		require.True(t, pacherr.IsNotExist(err))
+	})
+
+	t.Run("TestEmptyWrite", func(t *testing.T) {
+		t.Parallel()
+		client := newClient(t)
+		doWriteTest(t, client, tu.UniqueString("test-empty-write-"), []byte{})
+	})
+
+	t.Run("TestSingleWrite", func(t *testing.T) {
+		t.Parallel()
+		client := newClient(t)
+		doWriteTest(t, client, tu.UniqueString("test-single-write-"), []byte("foo bar"))
+	})
+
+	t.Run("TestSubdirectory", func(t *testing.T) {
+		t.Parallel()
+		client := newClient(t)
+		object := path.Join(tu.UniqueString("test-subdirectory-"), "object")
+		doWriteTest(t, client, object, []byte("foo bar"))
+	})
+
+	t.Run("TestIntegrity", func(t *testing.T) {
+		t.Parallel()
+		client := newClient(t)
+		name := "prefix/test-object"
+		expectedData, err := ioutil.ReadAll(io.LimitReader(rand.Reader, 1<<20))
+		require.NoError(t, err)
+		expectedHash := pachhash.Sum(expectedData)
+		require.NoError(t, client.Put(ctx, name, bytes.NewReader(expectedData)))
+		buf := &bytes.Buffer{}
+		require.NoError(t, client.Get(ctx, name, 0, 0, buf))
+		actualHash := pachhash.Sum(buf.Bytes())
+		require.Equal(t, expectedHash, actualHash)
+	})
+}
+
+// TestStorage is a defensive method for checking to make sure that storage is
+// properly configured.
+func TestStorage(ctx context.Context, c Client) error {
```

Comment:
We should probably run this during Pachd start up.

Reply:
That has the potential to litter object storage with undeleted objects.  If object storage doesn't work, I'm sure we will find out about it soon enough.  What is the advantage of failing startup? Wouldn't you want to keep retrying and troubleshoot other things without having to restart?

Reply:
Adding some sort of integrity checks to the storage layer, tracked and cleaned up by the usual mechanisms, might be worth doing.

Reply:
We want to catch object storage issues as soon as possible and in a place where it is obvious what the root problem is. I think a preflight check for object storage makes sense, but I don't have a strong opinion about where it occurs. Having it in the storage layer and using the tracking mechanism to handle cleanup makes sense to me.